### ISSUE TITLE
folly::small_vector needs to become empty after being moved from

### ISSUE THREAD

**Issue by tudor** [facebook/folly#1245]
`folly::small_vector` doesn't become empty after being moved from. It should, just like `std::vector`.

The following test fails (`v.size()` is 1, not 0).

```
TEST(SmallVector, Move) {
    folly::small_vector<uint64_t, 3> v;
    v.push_back(1);

    auto vv = std::move(v);
    EXPECT_EQ(1, vv.size());
    EXPECT_EQ(0, v.size());
}
```

**Comment by brenoguim** [comment]
A moved-from object should be in a valid but unspecified state. A vector with one element is still a valid vector, right? Why is this a problem?
You can always clear after you move, if you need that behavior.

**Comment by yfeldblum** [comment]
@brenoguim -

> A moved-from object should be in a valid but unspecified state.

It is true that for every type, a moved-from object of that type must be in a valid state. It is true that the standard does not specify any specific state that a moved-from object must be in, only that it be in some but possibly any valid state. To illustrate: one thing that the standard could in theory have specified is that, for any type for which equality-comparison is enabled, in addition to being in some valid state, a moved-from object of that type must compare equal to a default-constructed object of that type. But the standard doesn't specify that. The standard merely refrains from dictating anything that must hold across the board beyond validity.

But it is also true that, for some given type, the semantics of being moved-from can be specified further beyond mere validity. This holds for some types in the standard, including for `std::vector`.

> Move constructor. [...] After the move, `other` is guaranteed to be `empty()`.
>
> https://en.cppreference.com/w/cpp/container/vector/vector

**Comment by brenoguim** [comment]
I'm aware of that. But why should folly::small_vector care to follow the same std::vector rule?

**Comment by yfeldblum** [comment]
A few reasons come to mind:
* For the same reason that `std::vector` does - unless there is some overriding factor, if it's good enough for `std::vector` then it's good enough for `folly::small_vector`.
* To imitate `std::vector` as an explicit goal, which is separate from the above.
* To simplify `folly::small_vector` - simplification can be about reducing surprise, which is related to but also separate from the above.

**Comment by brenoguim** [comment]
I'm not a Folly implementer, I just follow the project for fun, but allow me to speculate:

If the small_vector is implemented using a buffer on the stack to store the N first elements, then the move operation (when size() < N) is probably implemented by simply moving all objects from the vector into the receiving vector.

So that's already something very different from std::vector. With std::vector, only pointers are changed because the data is stored on the heap.

If one wanted the small_vector to be empty afterwards, then all elements from the original vector would have to be destroyed, and size set to 0.

That might also be surprising, because that's something std::vector won't do.

Destroying all elements on the original vector is a needless, potentially expensive operation. And if you are using small_vector you might already have performance in mind.

Of course, if the type is trivial you could do that for almost free (just need to set size to 0), but it would incur in a difference from non-trivial to trivial types. That could also be surprising.

>For the same reason that std::vector does - unless there is some overriding factor, if it's good enough for std::vector then it's good enough for folly::small_vector.

The above would be the overriding factor I guess.

**Comment by tudor** [comment]
> I'm not a Folly implementer, I just follow the project for fun, but allow me to speculate:

**Comment by tudor** [comment]
Those elements have to be destroyed at some point. Either when you move, or when you destroy `other`.

And compatibility with `std::vector` is a huge reason -- I want to be able to replace `std::vector` with `folly::small_vector`, and the only side effect should be that I end up doing no memory allocations for the small-data use case :)

**Comment by brenoguim** [comment]
> and the only side effect should be that I end up doing no memory allocations for the small-data use case :)

Well, you will certainly get more side effects. If my speculation is right, you will get moves for all vector elements when `size() < N`, while `std::vector` will only fiddle with pointers.
`std::vector` will also guarantee pointer/iterator stability after moving as far as I recall, while a vector stored in the stack can't offer that.

> Those elements have to be destroyed at some point. Either when you move, or when you destroy other.

You can the use non-destroyed elements for other purposes, perhaps even re-filling them:

```
folly::small_vector<std::set<int>> v;
v.push_back(std::set<int>{1,2,3});

auto otherV = std::move(v);

v.back().clear();
v.back().insert(1); // I don't have to re-create a std::set here, I just reuse the one I had.
```

**Comment by yfeldblum** [comment]
Regarding reference stability, it must fundamentally be anticipated by users of any container type with in-situ storage that reference stability is never guaranteed, even when the type mimics another type without in-situ storage. This is a well-known price to pay for in-situ storage.

Regarding extra constructions and destructions on move, it must fundamentally be anticipated by users of any container type with in-situ storage that extra constructions and destructions may happen on move, even when the type mimics another type without in-situ storage. This is a well-known price to pay for in-situ storage.

Regarding what to do, I am not sure - there are arguments in both directions.

As one take, why not both, determined by policy? Seems like the default policy could be to clear the vector on move for compatibility with `std::vector` and overall lack of surprise. But an override policy could be not to clear the vector on move, for cases which specifically want to take advantage of that optimization.

**Comment by tudor** [comment]
> As one take, why not both, determined by policy? Seems like the default policy could be to clear the vector on move for compatibility with std::vector and overall lack of surprise. But an override policy could be not to clear the vector on move, for cases which specifically want to take advantage of that optimization.

Because that's punting the decision onto the user, and, in practice, the impact is likely to be small. I'd bet that most uses of `small_vector<T, N>` use a small value for `N`, in which case there aren't that many objects to destroy. Sure you can design a pathological benchmark that hypes up the cost of eager clear-on-move, but I question whether that's actually relevant.

I think the principle of least surprise is the overriding argument here. I replaced `std::vector` with `folly::small_vector` in a really trivial piece of code, and it stopped working, and it took me an hour to nail it down to this, so kill it :)

**Comment by brenoguim** [comment]
I agree the scenario I described should be very rare, and compatibility with std::vector sounds like a stronger point.

And to add more information to the thread:
`boost::container::small_vector` does not destroy the objects: https://godbolt.org/z/OvTYOq
`llvm::SmallVector` destroys the objects: https://godbolt.org/z/VWSj4Z

**Comment by ot** [comment]
I believe that this was an oversight, not a deliberate design decision. The proof is that move-assignment clears its argument (no matter whether inline or heap-allocated), only move-construction does not for inline vectors.

This was fixed in cb36f3c84958de88abf2c029950e5a2effdd1df4 .

I'm not a maintainer on github, but I think we can close this.
